R/qenv-join.R

```diff
@@ -1,12 +1,101 @@
 #' Join two `qenv` objects
 #'
-#' Combine two `qenv` objects by merging their environments and the code.
-#' Not all `qenv` objects can be combined:
-#' - if their environments contains objects of the same name but not identical
-#' - if `y` has unique code element placed before common element. This means that `y`
-#' in the environment of the `y` was evaluated some extra code before which can influence
-#' reproducibility
-#' - more cases to be done
+#' `join()` perform checks and merges two `qenv` objects into one `qenv` object.
+#'
+#' There are some situations where `join()` cannot be performed, such as:
+#' \itemize{
+#'   \item Both `qenv` objects contain an object of the same name but are not identical. \cr\cr
+#'   Example:
+#'   \preformatted{
+#'   x <- new_qenv(
+#'     code = c(mtcars1 = "mtcars1 <- mtcars"),
+#'     env = list2env(list(mtcars1 = mtcars))
+#'   )
+#'   y <- new_qenv(
+#'     code = c(mtcars1 = "mtcars1 <- mtcars['wt']"),
+#'     env = list2env(list(mtcars1 = mtcars['wt']))
+#'   )
+#'   z <- join(x, y)
+#'   # Error message will occur
+#'   }
+#'   In this example, `mtcars1` object exists in both `x` and `y` objects but the content are not identical.\cr
+#'   `mtcars1` in the `x qenv` object has more columns than `mtcars1` in the `y qenv` object (only has one column).
+#'   \item `join()` will look for identical `@id` values in both `qenv` objects.
+#'   The index position of these `@id`s must be the same to determine the evaluation order.
+#'   Otherwise, `join()` will throw an error message.\cr\cr
+#'   Example:
+#'   \preformatted{
+#'   common_q <- new_qenv(code = "v <- 1", env = list2env(list(v = 1)))
+#'   x <- eval_code(
+#'     common_q,
+#'     "x <- v"
+#'   )
+#'   y <- eval_code(
+#'     common_q,
+#'     "y <- v"
+#'   )
+#'   z <- eval_code(
+#'     y,
+#'     "z <- v"
+#'   )
+#'   q <- join(x, y)
+#'   join_q <- join(q, z)
+#'   # Error message will occur
+#'
+#'   # Check the order of evaluation based on the id slot
+#'   shared_ids <- intersect(q@id, z@id)
+#'   match(shared_ids, q@id) # Output: 1 3
+#'   match(shared_ids, z@id) # Output: 1 2
+#'   }
+#'   The error occurs because the index position of identical `@id` between the two objects is not the same.
+#'   \item The usage of temporary variable in the code expression could cause `join()` to fail. \cr\cr
+#'   Example:
+#'   \preformatted{
+#'   common_q <- new_qenv()
+#'   x <- eval_code(
+#'     common_q,
+#'     "x <- numeric(0)
+#'      for (i in 1:2) {
+#'        x <- c(x, i)
+#'      }"
+#'   )
+#'   y <- eval_code(
+#'     common_q,
+#'     "y <- numeric(0)
+#'      for (i in 1:3) {
+#'        y <- c(y, i)
+#'      }"
+#'   )
+#'   # Error message will occur
+#'
+#'   # Check the value of temporary variable i in both objects
+#'   x@env$i # Output: 2
+#'   y@env$i # Output: 3
+#'   }
+#'   `join()` fails because the temporary variable `i` exists in both objects but has different value.\cr
```

Comment:
```suggestion
#'   `join()` fails to provide a proper result because of the temporary variable `i` exists in both objects but has different value.\cr
```